File: Samples/Digital signatures/SignWithAwsKms/VB.NET/AwsSigner.vb
Imports System.IO
Imports Amazon.KeyManagementService
Imports Amazon.KeyManagementService.Model
Imports BitMiracle.Docotic.Pdf

Class AwsSigner
    Implements IPdfSigner

    Private ReadOnly m_keyId As String
    Private ReadOnly m_signingAlgorithm As SigningAlgorithmSpec

    Public Sub New(keyId As String, signingAlgorithm As SigningAlgorithmSpec)
        If signingAlgorithm = SigningAlgorithmSpec.SM2DSA Then
            Throw New ArgumentException("SM2DSA is not supported", NameOf(signingAlgorithm))
        End If

        m_keyId = keyId
        m_signingAlgorithm = signingAlgorithm
    End Sub

    Public ReadOnly Property SignatureAlgorithm As PdfSignatureAlgorithm Implements IPdfSigner.SignatureAlgorithm
        Get
            If m_signingAlgorithm = SigningAlgorithmSpec.ECDSA_SHA_256 OrElse
                m_signingAlgorithm = SigningAlgorithmSpec.ECDSA_SHA_384 OrElse
                m_signingAlgorithm = SigningAlgorithmSpec.ECDSA_SHA_512 Then
                Return PdfSignatureAlgorithm.Ecdsa
            End If

            If m_signingAlgorithm = SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256 OrElse
                m_signingAlgorithm = SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_384 OrElse
                m_signingAlgorithm = SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_512 Then
                Return PdfSignatureAlgorithm.Rsa
            End If

            If m_signingAlgorithm = SigningAlgorithmSpec.RSASSA_PSS_SHA_256 OrElse
                m_signingAlgorithm = SigningAlgorithmSpec.RSASSA_PSS_SHA_384 OrElse
                m_signingAlgorithm = SigningAlgorithmSpec.RSASSA_PSS_SHA_512 Then
                Return PdfSignatureAlgorithm.RsaSsaPss
            End If

            Throw New InvalidOperationException($"Unsupported {NameOf(SigningAlgorithmSpec)} value: {m_signingAlgorithm}")
        End Get
    End Property

    Public ReadOnly Property DigestAlgorithm As PdfDigestAlgorithm
        Get
            Dim alg As String = m_signingAlgorithm.Value
            If alg.EndsWith("256") Then Return PdfDigestAlgorithm.Sha256
            If alg.EndsWith("384") Then Return PdfDigestAlgorithm.Sha384
            If alg.EndsWith("512") Then Return PdfDigestAlgorithm.Sha512

            Throw New InvalidOperationException($"Unsupported {NameOf(SigningAlgorithmSpec)} value: {m_signingAlgorithm}")
        End Get
    End Property

    Public Function Sign(message As Byte()) As Byte() Implements IPdfSigner.Sign
        Using kmsClient = New AmazonKeyManagementServiceClient()
            Using stream = New MemoryStream(message)
                Dim signRequest = New SignRequest() With {
                    .SigningAlgorithm = m_signingAlgorithm,
                    .KeyId = m_keyId,
                    .MessageType = MessageType.RAW,
                    .Message = stream
                }
                Dim signResponse As SignResponse = kmsClient.SignAsync(signRequest).GetAwaiter().GetResult()
                Return signResponse.Signature.ToArray()
            End Using
        End Using
    End Function
End Class

File: Samples/Digital signatures/SignWithAwsKms/C#/AwsSigner.cs
using System;
using System.IO;
using Amazon.KeyManagementService;
using Amazon.KeyManagementService.Model;

namespace BitMiracle.Docotic.Pdf.Samples
{
    class AwsSigner : IPdfSigner
    {
        private readonly string m_keyId;
        private readonly SigningAlgorithmSpec m_signingAlgorithm;

        public AwsSigner(string keyId, SigningAlgorithmSpec signingAlgorithm)
        {
            if (signingAlgorithm == SigningAlgorithmSpec.SM2DSA)
                throw new ArgumentException("SM2DSA is not supported", nameof(signingAlgorithm));

            m_keyId = keyId;
            m_signingAlgorithm = signingAlgorithm;
        }

        public PdfSignatureAlgorithm SignatureAlgorithm
        {
            get
            {
                if (m_signingAlgorithm == SigningAlgorithmSpec.ECDSA_SHA_256 ||
                    m_signingAlgorithm == SigningAlgorithmSpec.ECDSA_SHA_384 ||
                    m_signingAlgorithm == SigningAlgorithmSpec.ECDSA_SHA_512)
                    return PdfSignatureAlgorithm.Ecdsa;

                if (m_signingAlgorithm == SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256 ||
                    m_signingAlgorithm == SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_384 ||
                    m_signingAlgorithm == SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_512)
                    return PdfSignatureAlgorithm.Rsa;

                if (m_signingAlgorithm == SigningAlgorithmSpec.RSASSA_PSS_SHA_256 ||
                    m_signingAlgorithm == SigningAlgorithmSpec.RSASSA_PSS_SHA_384 ||
                    m_signingAlgorithm == SigningAlgorithmSpec.RSASSA_PSS_SHA_512)
                    return PdfSignatureAlgorithm.RsaSsaPss;

                throw new InvalidOperationException($"Unsupported {nameof(SigningAlgorithmSpec)} value: {m_signingAlgorithm}");
            }
        }

        public PdfDigestAlgorithm DigestAlgorithm
        {
            get
            {
                string alg = m_signingAlgorithm.Value;
                if (alg.EndsWith("256"))
                    return PdfDigestAlgorithm.Sha256;

                if (alg.EndsWith("384"))
                    return PdfDigestAlgorithm.Sha384;

                if (alg.EndsWith("512"))
                    return PdfDigestAlgorithm.Sha512;

                throw new InvalidOperationException($"Unsupported {nameof(SigningAlgorithmSpec)} value: {m_signingAlgorithm}");
            }
        }

        public byte[] Sign(byte[] message)
        {
            using var kmsClient = new AmazonKeyManagementServiceClient();
            using var stream = new MemoryStream(message);
            var signRequest = new SignRequest()
            {
                SigningAlgorithm = m_signingAlgorithm,
                KeyId = m_keyId,
                MessageType = MessageType.RAW,
                Message = stream,
            };
            SignResponse signResponse = kmsClient.SignAsync(signRequest).GetAwaiter().GetResult();
            return signResponse.Signature.ToArray();
        }
    }
}
